### PR TITLE
cmd: Make `jj restore` (without `--from`) work if `@` is a merge commit 

### PR DESCRIPTION
**Pull request by ilyagr**
Fixes https://github.com/martinvonz/jj/issues/1228

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- (not planned) I have updated the documentation (README.md, docs/, demos/)
- (n/a) I have updated the config schema (src/config-schema.json)
- [x] I have added tests to cover my changes

### CODE REVIEW

File: tests/test_restore_command.rs [hunk 1]
```diff
@@ -100,3 +102,134 @@ fn test_restore() {
...
+    [38;5;1m   7[39m     : [38;5;1m>>>>>>>[39m
+         [38;5;2m   1[39m: [38;5;2mresolution[39m
+    "###);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["restore", "file"]);
+    insta::assert_snapshot!(stdout, @r###"
+    Created 70d64d389989 conflict
+    Working copy now at: 70d64d389989 conflict
+    Added 0 files, modified 1 files, removed 0 files
+    "###);
+    // Minor BUG: Restored conflict is flipped
```
Comment (ilyagr) [inline tests/test_restore_command.rs:157]: @martinvonz I'm not sure what's going on here. I checked, and this seems unrelated to your recent changes: rebasing this branch onto 3dfedf5 doesn't seem to change the behavior.

Comment (ilyagr) [reply tests/test_restore_command.rs:157]: Actually, maybe if I get the parents without resolving a revision, it'll go better?

Comment (ilyagr) [reply tests/test_restore_command.rs]: It did, this is now fixed.

**Comment by martinvonz** [comment]
> There is a minor bug that restoring can flip a conflict. I think this is an uncovered bug in conflict materialization, though I'm unsure.

Is that still true or did you figure it out?

**Comment by ilyagr** [comment]
> Is that still true or did you figure it out?

I figured it out and just fixed the commit description.

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 * `jj git push --deleted` will remove all locally deleted branches from the remote.
 
+* `jj restore` without `--from` works correctly even if `@` is a merge
+  commit. For now, there is an outstanding bug that conflicts may be restored
+  into a conflict that's presented differently than before (though the conflict
+  remains equivalent to what it was).
```
Comment (martinvonz) [inline CHANGELOG.md]: The bit about an outstanding bug is stale, right?

File: src/commands/mod.rs [hunk 3]
```diff
@@ -2490,24 +2490,34 @@ fn cmd_restore(
     args: &RestoreArgs,
 ) -> Result<(), CommandError> {
     let mut workspace_command = command.workspace_helper(ui)?;
+    let mut rebasing_from_parents = false; // The ways of using `Option` I can think of make this uglier.
```
Comment (martinvonz) [inline src/commands/mod.rs:2493]: I don't quite follow the comment. Do you mean that the way we currently use `Option` makes this uglier than it has to be?

Either way, do you think we can instead extract the code from `cmd_diff()`?  That code currently calculates `from_tree` and a `to_tree`, but it could easily calculate a `from_tree` and a `to_commit` instead.

File: CHANGELOG.md (hunk 2 above)
Comment (ilyagr) [reply CHANGELOG.md]: Yep. It's gone now.

File: src/commands/mod.rs (hunk 3 above)
Comment (ilyagr) [reply src/commands/mod.rs:2493]: Never mind the comment. The snippet from cmd_diff is much better.

Comment (ilyagr) [reply src/commands/mod.rs]: Done.

File: tests/test_restore_command.rs [hunk 4]
```diff
@@ -100,3 +102,124 @@ fn test_restore() {
...
+    } else {
+        let mut args = vec!["new", "-m", name];
+        args.extend(parents);
+        test_env.jj_cmd_success(repo_path, &args);
+    }
+    for (name, content) in files {
+        std::fs::write(repo_path.join(name), content).unwrap();
+    }
+    test_env.jj_cmd_success(repo_path, &["branch", "create", name]);
+}
```
Comment (martinvonz) [inline tests/test_restore_command.rs:221]: This looks familiar :) We should put it in `tests/common/mod.rs` (or similar), but that can be separate from this PR.

File: tests/test_restore_command.rs [hunk 5]
```diff
@@ -100,3 +102,124 @@ fn test_restore() {
...
+       1     : <<<<<<<
+       2     : %%%%%%%
+       3     : -base
+       4     : +a
+       5     : +++++++
+       6     : b
+       7     : >>>>>>>
+            1: resolution
+    "###);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["restore", "file"]);
```
Comment (martinvonz) [inline tests/test_restore_command.rs:151]: nit: Add at least a blank line here to mark that this is the end of setup. A comment saying something like "// Restore the the conflicted state" might also be useful

**Review by martinvonz — APPROVED** [review]

File: tests/test_restore_command.rs (hunk 4 above)
Comment (ilyagr) [reply tests/test_restore_command.rs:221]: Yeah. I also use a slightly different version of it in https://github.com/martinvonz/jj/pull/1146. I'm not sure if there's a way to have one version that works everywhere.

In any case, I won't do it here.

As usual, thanks for the review!

File: tests/test_restore_command.rs (hunk 5 above)
Comment (ilyagr) [reply tests/test_restore_command.rs:151]: This is helpful, thanks! I have written some huge tests recently, and I'm not always sure how to make them readable.
